crates/jit/src/instantiate.rs

```diff
@@ -517,19 +542,80 @@ impl CompiledModule {
         &self.text()[loc.start as usize..][..loc.length as usize]
     }
 
-    /// Returns the per-signature trampolines for this module.
-    pub fn trampolines(&self) -> impl Iterator<Item = (SignatureIndex, VMTrampoline, usize)> + '_ {
-        let text = self.text();
-        self.trampolines.iter().map(move |(signature, loc)| {
-            (
-                *signature,
-                unsafe {
-                    let ptr = &text[loc.start as usize];
-                    std::mem::transmute::<*const u8, VMTrampoline>(ptr)
-                },
-                loc.length as usize,
-            )
-        })
+    /// Get the array-to-Wasm trampoline for the function `index` points to.
+    ///
+    /// If the function `index` points to does not escape, then `None` is
+    /// returned.
+    ///
+    /// These trampolines are used for array callers (e.g. `Func::new`)
+    /// calling Wasm callees.
+    pub fn array_to_wasm_trampoline(&self, index: DefinedFuncIndex) -> Option<&[u8]> {
+        let idx = self
+            .array_to_wasm_trampolines
+            .binary_search_by_key(&index, |entry| entry.0)
+            .ok()?;
+        let (_, loc) = self.array_to_wasm_trampolines[idx];
+        Some(&self.text()[loc.start as usize..][..loc.length as usize])
+    }
+
+    /// Returns an iterator over all array-to-Wasm trampolines defined within
+    /// this module, providing both their index and their in-memory body.
+    pub fn array_to_wasm_trampolines(
+        &self,
+    ) -> impl ExactSizeIterator<Item = (DefinedFuncIndex, &[u8])> + '_ {
+        self.array_to_wasm_trampolines
+            .iter()
+            .map(move |(i, _)| (*i, self.array_to_wasm_trampoline(*i).unwrap()))
+    }
```

Comment:
I think the only use case for this is the profilers, and they don't have a great use for this. The profilers already have a lot of duplication of looking at all of these trampolines which isn't great, too.

This is fine to defer to another PR, but I wanted to write this down: these should all get removed and all the profilers should switch to using [this logic](https://github.com/bytecodealliance/wasmtime/blob/43ec4816f1280086987a6743d06368a606406bea/crates/wasmtime/src/trampoline/func.rs#L74-L103) which generically reads an ELF file and registers everything. There's specific support for DWARF in the jitdump profiler but I think that's all best removed at this point.

Again, future PR, just want to leave a reminder to myself to do this or open an issue later.

Reply:
I'll do this in a follow up.